### PR TITLE
HabitTracker App finished

### PR DESCRIPTION
**Pull request by DreamFXX**
Everything is written in my ReadMe file. Thanks for everything, i had plenty of fun doing this one!

Have a good day!

DreamFXX

### CODE REVIEW

**Review by DreamFXX — COMMENTED** [review]
codacy requirements fixed

**Comment by DreamFXX** [comment]
Completed challenges.

File: HabitTracker.ConsoleApp-DreamFXX/HabitTrackerPersonal-ConsoleApp.db
Comment (chrisjamiecarter) [inline HabitTracker.ConsoleApp-DreamFXX/HabitTrackerPersonal-ConsoleApp.db:1]: 🟠 Database files shouldn't be included in source control.

💡 They are ever changing, so will always flag as changed, they can grow quite large in size, and may contain data you do not want shared.

🔧 Add to the .gitignore file.

File: README.md
Comment (chrisjamiecarter) [inline README.md:1]: 🟢 Readme file

💡 Well done on your readme file, the contents are a perfect introduction to your application.

File: HabitTracker.ConsoleApp-DreamFXX/Properties/launchSettings.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "profiles": {
+    "HabitTracker_ConsoleApp": {
+      "commandName": "Project",
+      "workingDirectory": "C:\\Users\\marti\\source\\repos\\HabitTracker.ConsoleApp-DreamFXX",
```
Comment (chrisjamiecarter) [inline HabitTracker.ConsoleApp-DreamFXX/Properties/launchSettings.json:5]: 🟠 Hard-coded magic strings

💡 Remember that this path with exist on no one else's machine and therefore will need to be reconfigured.

🔧 A note in the README to highlight this prerequisite and to change this before launching the app.

File: HabitTracker.ConsoleApp-DreamFXX/Program.cs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+        using (var connection = new SQLiteConnection(connectionString))
+        {
+            connection.Open();
+            var tableCmd = connection.CreateCommand();
+            tableCmd.CommandText = "SELECT * FROM HabitRecords";
+
+            List<HabitRecord> tableData = new List<HabitRecord>();
+            SQLiteDataReader reader = tableCmd.ExecuteReader();
+
+            if (reader.HasRows)
```
Comment (chrisjamiecarter) [inline HabitTracker.ConsoleApp-DreamFXX/Program.cs:104]: 🟠 Code Complexity

💡 `if (reader.HasRows)` is a redundant level of indentation here, as `while (reader.Read())` will only be true if the reader has rows. Essentially, `reader.Read()` already does this check for you.

🔧 Remove the redundant `if` statement and only keep the `while` statement.

File: HabitTracker.ConsoleApp-DreamFXX/Program.cs [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+            {
+                Console.WriteLine("\n\nNo records were found!\n\n");
+            }
+
+            connection.Close();
+
+            Console.WriteLine("------------HABIT RECORDS-----------\n");
+            foreach (var record in tableData)
+            {
+                Console.WriteLine($"{record.Id} -> {record.Date} in {record.Time}h // {record.Quantity}");
```
Comment (chrisjamiecarter) [inline HabitTracker.ConsoleApp-DreamFXX/Program.cs:129]: 🟢 String Interpolation

File: HabitTracker.ConsoleApp-DreamFXX/Program.cs [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        int habitId = GetNumberInput("Habit ID:");
+        string date = GetDate();
+        string time = GetTime();
+        int quantity = GetNumberInput("Enter quantity of habit you consumed // ran // did in units you selected in specified habit tracking.");
+
+        using (var connection = new SQLiteConnection(connectionString))
+        {
+            connection.Open();
+            var tableCmd = connection.CreateCommand();
+            tableCmd.CommandText = $"INSERT INTO HabitRecords (HabitId, Date, Time, Quantity) VALUES ({habitId}, '{date}', '{time}', {quantity})";
```
Comment (chrisjamiecarter) [inline HabitTracker.ConsoleApp-DreamFXX/Program.cs:149]: 🟠 Parameterised SQL Queries

💡 I know this is technically a challenge requirement. But every query should be parameterised as it is not hard to refactor and the dangers of not doing so are great.

🔧 
❌
```
tableCmd.CommandText = $"INSERT INTO HabitRecords (HabitId, Date, Time, Quantity) VALUES ({habitId}, '{date}', '{time}', {quantity})";
```

✔️ 
```
tableCmd.CommandText = $"INSERT INTO HabitRecords (HabitId, Date, Time, Quantity) VALUES (@HabitId, @Date, @Time, @Quantity)";
command.Parameters.AddWithValue("@HabitId", habitId);
command.Parameters.AddWithValue("@Date", date);
command.Parameters.AddWithValue("@Time", time);
command.Parameters.AddWithValue("@Quantity", quantity);
```

File: HabitTracker.ConsoleApp-DreamFXX/Program.cs [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    internal static string GetDate()
+    {
+        Console.WriteLine("\n\nEnter a date. // Enter 0 to go back to the menu.\n\n");
+        Console.Write("Type the date in this order -> DD-MM-YYYY - ");
+
+        string dateInput = Console.ReadLine();
+
+        while (!DateTime.TryParseExact(dateInput, "dd-MM-yyyy", new CultureInfo("en-US"), DateTimeStyles.None, out _))
```
Comment (chrisjamiecarter) [inline HabitTracker.ConsoleApp-DreamFXX/Program.cs:239]: 🟢 `DateTime.TryParseExact`

File: HabitTracker.ConsoleApp-DreamFXX/Program.cs [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+                tableCmd.ExecuteNonQuery();
+            }
+            connection.Close();
+        }
+        Console.WriteLine("Testing records (100) were sucessfully created and added to their specified tables.");
+    }
+
+    // Properties class
+
+    public class HabitRecord
```
Comment (chrisjamiecarter) [inline HabitTracker.ConsoleApp-DreamFXX/Program.cs:357]: 🟠 Object Orientated Programming

💡 Again, I know the project scope allows for this. But I would really of liked to see this in it's own file `HabitRecord.cs`

🔧 Refactor into it's own file.

File: README.md
Comment (chrisjamiecarter) [inline README.md:1]: 🟠 File Location

💡 Currently this file exists outside of your project.

🔧 It should be moved inside you project folder.

**Review by chrisjamiecarter — APPROVED** [review]
Hey @DreamFXX 👋,

Firstly, excellent work on your HabitTracker project submission 🎉 !

I have performed a peer review, please see my individual comments for praise and suggestions 💬.

🟠 **Solution Structure**
Close! But as per the [Code Reviews](https://thecsharpacademy.com/article/52/code-reviews) academy page, your solution files should be contained inside a new folder named [Project].[GitHubUserName].
❌ 
`HabitTracker.ConsoleApp-DreamFXX`
✔️ 
`HabitTracker.DreamFXX`

🟠 **Style Guide Compliance**
Remember that when you have finished a project, give it a polish pass. Pay attention to any code conventions in place, in this instance refer to the C# Academy's [code-conventions](https://thecsharpacademy.com/article/58/code-conventions) page.
This is important as good code conventions/styling ensures consistency, makes your code easier for others to read and maintain, and improves your chances of getting a pull request reviewed and approved.

🟢 & 🟠 **Data Seeding**
Fantastic job on implementing a data seeding method in your app. However it does input 100 records on startup. Refactoring to only seed on a fresh database or program it to a menu option would improve the user experience.

🟠  **Inconsistent Data Format**
Seeded data is in format `dd-MM-yy`, input first asks for format `dd-MM-yyyy` (albeit `DD-MM-YYYY`), then for `dd-mm-yy`.

image

image

🟠 **Input Validation**
I can input 3 when only 1 is available

image

I can type gibberish as a time value

image

***

I suggest revisiting to try understand and fix any issues 🔧 , but as these relate to the extra challenge requirement they will not block the PR.

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points 🆘 .

You're doing a fantastic job, and these are just a few suggestions to support your growth as a developer.

I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter 👍

**Comment by DreamFXX** [comment]
Fixed all your issues. Thank you so much for honest review, helped me a lot!

-DreamFXX

**Pull request closed without merging**